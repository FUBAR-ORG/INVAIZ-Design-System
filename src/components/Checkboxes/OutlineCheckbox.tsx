import type { CheckboxDefaultProps } from '@components/Checkboxes/interfaces/Checkbox.interface';
// types

import { type ReactNode, useMemo } from 'react';
// React modules

import {
  CLASSNAME_OUTLINE_CHECK_ICON,
  StyleLabel,
  StyleOutlineCheckIcon,
  StyleOutlineCheckbox,
  StyleOutlineCheckboxInput,
  StyleOutlineCheckboxHover,
  StyleCheckboxText,
} from '@components/Checkboxes/styles/OutlineCheckbox.style';
// styles

interface OutlineCheckboxBaseProps extends OutlineCheckboxProps {
  /**
   * 내부에 렌더링할 아이콘 컴포넌트
   */
  iconComponent: ReactNode;
}

/**
 * 가장 자리가 강조된 체크박스의 베이스 컴포넌트입니다.
 */
const OutlineCheckboxBase = ({
  iconComponent,
  text,
  onChange,
  disabled,
  ...props
}: OutlineCheckboxBaseProps) => (
  <StyleLabel>
    <StyleOutlineCheckboxHover disabled={disabled}>
      <StyleOutlineCheckboxInput
        type='checkbox'
        onChange={(e) => onChange?.(e.target.checked)}
        disabled={disabled}
        {...props}
      />
      <StyleOutlineCheckbox>{iconComponent}</StyleOutlineCheckbox>
    </StyleOutlineCheckboxHover>
    <StyleCheckboxText disabled={disabled}>{text}</StyleCheckboxText>
  </StyleLabel>
);

interface OutlineCheckboxProps extends CheckboxDefaultProps {
  /**
   * 체크박스 우측에 표시될 텍스트입니다.
   */
  text: string;
}

/**
 * 가장 자리 테두리가 강조된 체크박스입니다.
 */
const OutlineCheckbox = ({ ...props }: OutlineCheckboxProps) => (
  <OutlineCheckboxBase
    iconComponent={<StyleOutlineCheckIcon size={20} icon='Check' />}
    {...props}
  />
);

interface ParentCheckboxProps extends OutlineCheckboxProps {
  /**
   * 불확실한 상태 유무입니다.
   */
  isIndeterminate?: boolean;
}

/**
 * 서브 체크박스 리스트의 부모 역할 체크박스입니다.
 */
export const ParentCheckbox = ({ isIndeterminate, ...props }: ParentCheckboxProps) => {
  const indeterminateIcon = useMemo(
    () =>
      isIndeterminate ? (
        <StyleOutlineCheckIcon icon='Subtract' size={20} className={CLASSNAME_OUTLINE_CHECK_ICON} />
      ) : (
        <StyleOutlineCheckIcon icon='Check' size={20} />
      ),
    [isIndeterminate]
  );

  return <OutlineCheckboxBase iconComponent={indeterminateIcon} {...props} />;
};

export default OutlineCheckbox;
