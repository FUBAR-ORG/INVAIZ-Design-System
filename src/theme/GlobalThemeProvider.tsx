import { PropsWithChildren } from 'react';
import { ThemeProvider } from '@emotion/react';
import font, { fontSize, fontWeight, lineHeight } from '@theme/font';
import color from '@theme/color';

export default function GlobalThemeProvider({ children }: PropsWithChildren<unknown>) {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>;
}

const theme = { font, fontSize, lineHeight, fontWeight, color };
