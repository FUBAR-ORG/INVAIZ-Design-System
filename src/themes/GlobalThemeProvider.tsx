import type { PropsWithChildren } from 'react';
import type { EmotionTheme } from '@themes/emotion-theme';
import type { ColorTheme } from '@themes/colors/color';
// types

import { ThemeProvider } from '@emotion/react';
// React modules

import font, { fontSize, fontWeight, lineHeight, textAlign, fontColor } from '@themes/font';
import { colorMap, normal, light } from '@themes/colors/color';
import style from '@themes/style';
import GlobalStyle from '@themes/GlobalStyle';

interface GlobalThemeProviderProps {
  colorTheme?: ColorTheme;
}

export default function GlobalThemeProvider({
  children,
  colorTheme = 'normal',
}: PropsWithChildren<GlobalThemeProviderProps>) {
  const theme: EmotionTheme = {
    normal,
    light,

    color: colorMap[colorTheme],

    font,
    fontSize,
    lineHeight,
    fontWeight,
    textAlign,
    fontColor,

    style,
  };
  return (
    <ThemeProvider theme={theme}>
      {children}
      <GlobalStyle />
    </ThemeProvider>
  );
}
