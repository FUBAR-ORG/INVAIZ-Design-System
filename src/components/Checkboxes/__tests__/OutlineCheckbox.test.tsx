import { render } from '@tests/test-utils';
import { create } from 'react-test-renderer';
// test utils

import OutlineCheckbox from '@components/Checkboxes/OutlineCheckbox';
// components

import GlobalThemeProvider from '@themes/GlobalThemeProvider';
// providers

const OUTLINE_TEXT = 'Outline Checkbox' as const;
// constants

describe('OutlineCheckbox', () => {
  it('체크박스의 기본값은 `false`이며, 변경이 가능하다.', () => {
    const { getAllByRole } = render(
      <>
        <OutlineCheckbox text={OUTLINE_TEXT} />
        <OutlineCheckbox text={OUTLINE_TEXT} checked />
      </>
    );
    const [isNotChanged, isChanged] = getAllByRole('checkbox');

    expect(isNotChanged).not.toBeChecked();
    expect(isChanged).toBeChecked();
  });

  it('체크박스는 클릭할 수 있으며, 클릭 시 `checked` 상태 변경이 가능하다.', () => {
    const onChange = jest.fn();
    const { getByRole } = render(<OutlineCheckbox text={OUTLINE_TEXT} onChange={onChange} />);
    const checkbox = getByRole('checkbox');

    expect(onChange).not.toHaveBeenCalled();

    checkbox.click();
    expect(onChange).toHaveBeenCalledWith(true);
    expect(checkbox).toBeChecked();

    checkbox.click();
    expect(onChange).toHaveBeenCalledWith(false);
    expect(checkbox).not.toBeChecked();

    expect(onChange).toHaveBeenCalled();
  });

  it('체크박스는 텍스트 영역을 클릭하여도 상태 변경이 가능하다.', () => {
    const { getByRole, getByText } = render(<OutlineCheckbox text={OUTLINE_TEXT} />);

    const checkbox = getByRole('checkbox');
    const textArea = getByText(OUTLINE_TEXT);

    textArea.click();
    expect(checkbox).toBeChecked();
  });

  it('체크박스는 `disabled`일 때, 상태 변경이 불가능하다.', () => {
    const { getByRole } = render(<OutlineCheckbox text={OUTLINE_TEXT} disabled />);
    const checkbox = getByRole('checkbox');

    checkbox.click();
    expect(checkbox).not.toBeChecked();
  });
});

test('OutlineCheckbox Snapshot', () => {
  const checked = create(
    <GlobalThemeProvider>
      <OutlineCheckbox text={OUTLINE_TEXT} checked />
    </GlobalThemeProvider>
  );
  expect(checked).toMatchSnapshot();

  const unChecked = create(
    <GlobalThemeProvider>
      <OutlineCheckbox text={OUTLINE_TEXT} />
    </GlobalThemeProvider>
  );
  expect(unChecked).toMatchSnapshot();
});
