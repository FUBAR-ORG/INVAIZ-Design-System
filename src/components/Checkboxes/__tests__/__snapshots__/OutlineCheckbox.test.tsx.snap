// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutlineCheckbox Snapshot 1`] = `
<label
  className="css-0"
>
  <span
    className="css-1u8zmlt"
  >
    <input
      checked={true}
      className="css-18gtr08"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="css-3xhjcj"
    >
      <svg
        className="css-bw5rgr"
        data-file-name="Check.svg"
        data-svg-name="Check"
        data-test-id="Check"
        height={20}
        width={20}
      />
    </div>
  </span>
  <p
    className="css-1m9p5pv"
  >
    Outline Checkbox
  </p>
</label>
`;

exports[`OutlineCheckbox Snapshot 2`] = `
<label
  className="css-0"
>
  <span
    className="css-1u8zmlt"
  >
    <input
      className="css-18gtr08"
      onChange={[Function]}
      type="checkbox"
    />
    <div
      className="css-3xhjcj"
    >
      <svg
        className="css-bw5rgr"
        data-file-name="Check.svg"
        data-svg-name="Check"
        data-test-id="Check"
        height={20}
        width={20}
      />
    </div>
  </span>
  <p
    className="css-1m9p5pv"
  >
    Outline Checkbox
  </p>
</label>
`;
